### PR TITLE
[Notifier ] Add Discord notifier

### PR DESCRIPTION
**Pull request by mpiot**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#13558

Add a basic Discord notifier via webhook (https://discordapp.com/developers/docs/resources/webhook).

Capture d’écran de 2020-04-17 15-27-42

Actually I've just a Message from the notification displayed (on the image the simple line: "My notification subject").

It's also possible to have an embed message (https://discordapp.com/developers/docs/resources/channel#embed-object) as exemple I've done one with a title (message subject), description (message content), and a color (on the importance)

Capture d’écran de 2020-04-17 15-27-54

At the moment it's not implemented yet in this PR, 2 possibilities:
 - the embed type by default (never the simple type)
 - add a DiscordConfigurationMessage object  to enable the embed format ? (only on ChatMessage, not Notification)

In my case, I prefer embded by default, like Email notification (subject, content and border color).

We also can add a configuration file (only for ChatMessage) to the Bot name, bot image.

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+use Symfony\Component\Notifier\Exception\LogicException;
+use Symfony\Component\Notifier\Exception\TransportException;
+use Symfony\Component\Notifier\Message\ChatMessage;
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Transport\AbstractTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * DiscordTransport.
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:23]:
```suggestion

```

**Comment by walva** [comment]
@mpiot I'm willing to help! Is there something I can do? I'm currently trying to test your new feature.

**Comment by mpiot** [comment]
Hi @walva, actually I'm waiting some discussion about this implementation. In fact, actually I've implemented it very simply (the first image) and without any configuration.

I think I will add a Configuration to set the bot name and bot image.

A part to discuss is about the message format, in Discord Webhook api, we can sent a simple message as already done here, or as embeds (image 2) with a border, title, footer, description (the simplest possibility). Use it by default ? (border color with importance, title with notification subject, description with content).

Maybe other ideas ? I don't know.

**Comment by noniagriconomie** [comment]
Iirc there is a slack bridge, maybe do a « replication » of this bridge ?
As both are similar in main points

**Comment by mpiot** [comment]
Yes there is a SlackBridge,  but I don't know the Slack API, I think you want to speak about the Block part of the Slack Notifier ?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml [hunk 2]
```diff
@@ -46,6 +46,10 @@
             <tag name="texter.transport_factory" />
         </service>
 
+        <service id="notifier.transport_factory.discord" class="Symfony\Component\Notifier\Bridge\Discord\DiscordTransportFactory" parent="notifier.transport_factory.abstract">
+            <tag name="chatter.transport_factory" />
+        </service>
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml:51]: You should rebase and move this configuration to notifier_transports.php file.

File: src/Symfony/Component/Notifier/Bridge/Discord/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.1.0
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Discord/CHANGELOG.md:4]: 5.2

File: src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ *
+ * @author Mathieu Piot <math.piot@gmail.com>
+ *
+ * @internal
+ *
+ * @experimental in 5.0
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:28]: should be 5.2 in all new files.

File: src/Symfony/Component/Notifier/Bridge/Discord/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+Discord Notifier
+================
+
+Provides Discord integration for Symfony Notifier.
+
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Discord/README.md:5]: You should probably add the DSN pattern here.

File: src/Symfony/Component/Notifier/Bridge/Discord/composer.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": "^7.2.5",
+        "symfony/http-client": "^4.3|^5.0",
+        "symfony/notifier": "^5.1"
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Discord/composer.json:21]: 5.2

File: src/Symfony/Component/Notifier/Bridge/Discord/composer.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    "autoload": {
+        "psr-4": { "Symfony\\Component\\Notifier\\Bridge\\Discord\\": "" },
+        "exclude-from-classmap": [
+            "/Tests/"
+        ]
+    },
+    "minimum-stability": "dev",
+    "extra": {
+        "branch-alias": {
+            "dev-master": "5.1-dev"
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Discord/composer.json:35]: 5.2

File: src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    public function supports(MessageInterface $message): bool
+    {
+        return $message instanceof ChatMessage;
+    }
+
+    /**
+     * @see https://discordapp.com/developers/docs/resources/webhook
+     */
+    protected function doSend(MessageInterface $message): void
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:59]: This must now return a `SentMessage`.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by fabpot** [comment]
@mpiot Can you submit a PR on symfony/recipes (you can have a look at other notifier bridge recipes to get some inspiration)?

File: src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * @author Mathieu Piot <math.piot@gmail.com>
+ *
+ * @internal
+ *
+ * @experimental in 5.0
+ */
+final class DiscordTransport extends AbstractTransport
+{
+    protected const HOST = 'discordapp.com';
```
Comment (skmedix) [inline src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:32]:
```suggestion
    protected const HOST = 'discord.com';
```

File: src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        return sprintf('discord://%s?channel=%s', $this->getEndpoint(), $this->chatChannel);
+    }
+
+    public function supports(MessageInterface $message): bool
+    {
+        return $message instanceof ChatMessage;
+    }
+
+    /**
+     * @see https://discordapp.com/developers/docs/resources/webhook
```
Comment (skmedix) [inline src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:57]:
```suggestion
     * @see https://discord.com/developers/docs/resources/webhook
```

**Review by skmedix — CHANGES_REQUESTED** [review]

**Comment by mpiot** [comment]
@fabpot I look at this (changes and recipe) in few days (hollidays).
@skmedix Thanks, They've change URLs :( I'll fix it with the rest.

**Comment by fabpot** [comment]
@mpiot Friendly ping :)

**Comment by mpiot** [comment]
@fabpot Done for changes on this PR and open symfony/recipes#816

File: src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        $response = $this->client->request('POST', $endpoint, [
+            'json' => array_filter($options),
+        ]);
+
+        if (204 !== $response->getStatusCode()) {
+            $result = $response->toArray(false);
+
+            throw new TransportException(sprintf('Unable to post the Discord message: "%s" (%s).', $result['message'], $result['code']), $response);
+        }
+    }
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:76]: Missing the `return` statement... which makes me think that this should be tested before merging (at least manually).

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php (hunk 11 above)
Comment (mpiot) [reply src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:76]: Oh yes my bad... I've just replace the return statement :void to :SentMessage. I went too fast and didn't pay attention that it was necessary to  add a corresponding return statement in the function... :/

Comment (fabpot) [reply src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:76]: Are you still interested in finishing this PR?

Comment (mpiot) [reply src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:76]: I don't have much time lately, I'll try to watch this when I'm free.

Comment (fabpot) [reply src/Symfony/Component/Notifier/Bridge/Discord/DiscordTransport.php:76]: Thank you for the quick reply. I'm going to close then. Perhaps someone can take over to finish it.

**Comment by fabpot** [comment]
Anyone willing to take over this one. I think most of the work is already done. Apart from my last comment, this should also be tested in real life to be sure it works well.

**Comment by connorhu** [comment]
I will finish the work on this PR.

**Comment by mpiot** [comment]
@connorhu Thanks, I was going to try to watch. But I keep you see it. I've just try to link the code with php link to test on my project (then add the good return statement) but have an `Cannot dump definition because of invalid class name ('chatter.transport_factory').`  (#37298).

**Comment by mpiot** [comment]
@fabpot @connorhu By applying the patch #38175, I've can test my PR and it works well.

**Comment by connorhu** [comment]
I [finished](https://github.com/connorhu/symfony/commits/discord-logger-handler) the support of rich type webhook. Result of the [spotify test example](https://birdie0.github.io/discord-webhooks-guide/examples/spotify.html):
image

What's next?

Example:
```php
$chatter = new Chatter($discordTransport);
$chatMessage = new ChatMessage('');
$discordOptions = (new DiscordOptions())
    ->username('connor bot')
    ->addEmbed((new DiscordEmbed())
        ->color(2021216)
        ->title('New song added!')
        ->thumbnail((new DiscordMediaEmbedObject())
        ->url('https://i.scdn.co/image/ab67616d0000b2735eb27502aa5cb1b4c9db426b'))
        ->addField((new DiscordFieldEmbedObject())
            ->name('Track')
            ->value('[Common Ground](https://open.spotify.com/track/36TYfGWUhIRlVjM8TxGUK6)')
            ->inline(true)
        )
        ->addField((new DiscordFieldEmbedObject())
            ->name('Artist')
            ->value('Alasdair Fraser')
            ->inline(true)
        )
        ->addField((new DiscordFieldEmbedObject())
            ->name('Album')
            ->value('Dawn Dance')
            ->inline(true)
        )
        ->footer((new DiscordFooterEmbedObject())
            ->text('Added ...')
            ->iconUrl('https://upload.wikimedia.org/wikipedia/commons/thumb/1/19/Spotify_logo_without_text.svg/200px-Spotify_logo_without_text.svg.png')
        )
    )
;
$chatMessage->options($discordOptions);
$chatter->send($chatMessage);
```

**Comment by mpiot** [comment]
@connorhu Great job ;)
I think since this PR is closed, you can open a new one, if you want you can link this new one on:

Doc PR: symfony/symfony-docs#13558 (if you have the same organization)
Recipe PR: symfony/recipes#816

Both of this PRs should work, and maybe add a part in the documentation about the rich type ?

**Comment by fabpot** [comment]
@connorhu Do you want to open a new PR?

**Comment by mpiot** [comment]
@connorhu If you can’t say me. I’ll create it and fix doc and recipe PR. I wait before fix them.

**Comment by connorhu** [comment]
@fabpot Yes I will create a new PR, but first I need to create some tests for the new classes.

**Comment by mpiot** [comment]
@connorhu do you need some help for tests and open the new PR ? The 5.2 is really close.

**Comment by connorhu** [comment]
> @connorhu do you need some help for tests and open the new PR ? The 5.2 is really close.

I'm done the rest of the work. Nobody tested the code except me, so it's really experimental.

**Pull request closed without merging**